### PR TITLE
2226 - 2

### PR DESCRIPTION
**Pull request by andreydro**
## Номер
2226

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/6Sqz6TRSUiU

### CODE REVIEW

File: 2226/2/versus.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+require 'fuzzy_match'
+require 'optparse'
+# :reek:RepeatedConditional
+# :reek:TooManyStatements
+# :reek:UncommunicativeVariableName
+# :reek:UtilityFunction
+# :reek:FeatureEnvy
+# rubocop:disable Metrics/ClassLength
+class Raper
+  SWEAR_WORDS_ARRAY = [
```
Comment (stahor) [inline 2226/2/versus.rb]: %w

File: 2226/2/versus.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    'блядословник', 'блядский', 'блядство', 'въебаться', 'взбляд', 'впиздячил', 'выблядовал', 'выблядок',
+    'выебон', 'выёбывается', 'глупизди', 'доебался', 'ебанёшься', 'ебанул', 'ебашит', 'ёбнул', 'ебало',
+    'ебанулся', 'ебнул', 'жидоёб', 'жидоёбский', 'заёб', 'изъебнулся', 'заебал', 'заебись', 'козлоёб',
+    'козлоёбище', 'хуй', 'пизда', 'пиздец', '\*', 'сука'
+  ].freeze
+
+  SWEAR_WORDS = SWEAR_WORDS_ARRAY.join('|')
+
+  def filter_rapers_from_the_list(file_name)
+    first_filter = file_name.sub(/ротив_/, '')
```
Comment (stahor) [inline 2226/2/versus.rb]: why not `file_name.sub(...).sub(...)...`

File: 2226/2/versus.rb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    third_filter.sub(/_$/, '')
+  end
+
+  def list_all_rapers
+    raper_storage = []
+
+    Dir.glob('rap-battles/*') do |file_name|
+      filtered_data = filter_rapers_from_the_list(file_name)
+      raper_storage << filtered_data.split('_п')
+      raper_storage.flatten!.uniq!
```
Comment (stahor) [inline 2226/2/versus.rb]: do this one time outside loop

File: 2226/2/versus.rb [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    end
+    raper_storage
+  end
+
+  def number_of_battles(name)
+    battles_storage = [1]
```
Comment (stahor) [inline 2226/2/versus.rb]: ???

File: 2226/2/versus.rb [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    Dir.glob('rap-battles/*') do |file_name|
```
Comment (stahor) [inline 2226/2/versus.rb]: `each_with_object`

File: 2226/2/versus.rb [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      battles_storage << file_name.match(name)
+      battles_storage.reject!(&:nil?)
+    end
+    battles_storage.size / 2
+  end
+
+  def number_of_swear_words(name)
+    swear_words_storage = []
+
+    Dir.glob('rap-battles/*') do |file_name|
```
Comment (stahor) [inline 2226/2/versus.rb]: `each_with_object`

File: 2226/2/versus.rb [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+  def average_number_swearing_words_in_battle(name)
+    average = number_of_swear_words(name).fdiv(number_of_battles(name))
+    average.round(2)
+  end
+
+  def number_of_words_in_rounds(name)
+    number_of_words_in_rounds = 0
+
+    Dir.glob('rap-battles/*') do |file_name|
```
Comment (stahor) [inline 2226/2/versus.rb]: `each_with_object`

File: 2226/2/versus.rb [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    battles_storage.size / 2
+  end
+
+  def number_of_swear_words(name)
+    swear_words_storage = []
+
+    Dir.glob('rap-battles/*') do |file_name|
+      if file_name.match(name)
+        text = File.read(file_name)
+        swear_words_storage << text.scan(/#{SWEAR_WORDS}/)
```
Comment (stahor) [inline 2226/2/versus.rb]: +=

File: 2226/2/versus.rb [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    number_of_words_in_rounds
+  end
+
+  def number_of_rounds(name)
+    round_words = []
+
+    Dir.glob('rap-battles/*') do |file_name|
+      if file_name.match(name)
+        text = File.read(file_name)
+        round_words << text.match(/Раунд/)
```
Comment (stahor) [inline 2226/2/versus.rb]: sometimes round is not used in the file at all. And sometimes they just shout this word at the end of the round. So this won't work

File: 2226/2/versus.rb [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        round_words.reject!(&:nil?)
+      end
+    end
+    round_words
+  end
+
+  def average_number_words_in_round(name)
+    number_of_rounds = number_of_rounds(name).size
+    number_of_rounds = 1 if number_of_rounds.zero?
```
Comment (stahor) [inline 2226/2/versus.rb]: this should be in number_of_rounds method

File: 2226/2/versus.rb [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+    number_of_words_in_rounds(name) / number_of_rounds
+  end
+
+  def sort_and_reverse_hash(hash_for_sorting, sorted_storage)
+    hash_for_sorting = hash_for_sorting.sort_by { |_key, value| value[1] }
```
Comment (stahor) [inline 2226/2/versus.rb]: just _

File: 2226/2/versus.rb [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    hash_for_sorting.reverse!
```
Comment (stahor) [inline 2226/2/versus.rb]: sort by `- value [1]`

File: 2226/2/versus.rb [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    hash_for_sorting.each do |elem|
```
Comment (stahor) [inline 2226/2/versus.rb]: don't u want to sort storage instead?

File: 2226/2/versus.rb [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+      sorted_storage[elem[0]] = elem[1]
+    end
+    sorted_storage
+  end
+
+  def statistics(number)
+    unsorted_storage = {}
+    sorted_storage = {}
+    list_all_rapers.each do |raper|
```
Comment (stahor) [inline 2226/2/versus.rb]: each_with_object
and why not just `unsorted_storage[raper] << [number_of_battles(raper), number_of_swear_words(raper), ...]`

File: 2226/2/versus.rb [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+  def sort_and_reverse_hash(hash_for_sorting, sorted_storage)
+    hash_for_sorting = hash_for_sorting.sort_by { |_key, value| value[1] }
+    hash_for_sorting.reverse!
+    hash_for_sorting.each do |elem|
+      sorted_storage[elem[0]] = elem[1]
+    end
+    sorted_storage
+  end
+
+  def statistics(number)
```
Comment (stahor) [inline 2226/2/versus.rb]: number of what?

File: 2226/2/versus.rb [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+      storage_for_each_raper << number_of_battles(raper) << number_of_swear_words(raper)
+      storage_for_each_raper << average_number_swearing_words_in_battle(raper) << average_number_words_in_round(raper)
+      unsorted_storage[raper] = storage_for_each_raper
+    end
+    sorted_storage = sort_and_reverse_hash(unsorted_storage, sorted_storage)
+    sorted_storage.first(number.to_i)
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def make_table(number)
```
Comment (stahor) [inline 2226/2/versus.rb]: number of stars in the sky I guess

File: 2226/2/versus.rb [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+    filtered_data = {}
+    statistics(number).each do |elem|
+      filtered_data[elem[0]] = elem[1]
+    end
+    rows = []
+    filtered_data.each do |k, v|
+      rows << [k.to_s, v[0].to_s + ' баттлов', v[1].to_s + ' нецензурных слов',
```
Comment (stahor) [inline 2226/2/versus.rb]: break into more lines and use string interpolation

File: 2226/2/versus.rb [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+      unsorted_storage[raper] = storage_for_each_raper
+    end
+    sorted_storage = sort_and_reverse_hash(unsorted_storage, sorted_storage)
+    sorted_storage.first(number.to_i)
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def make_table(number)
+    filtered_data = {}
+    statistics(number).each do |elem|
```
Comment (stahor) [inline 2226/2/versus.rb]: do u really need this?

File: 2226/2/versus.rb [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  # rubocop:enable Metrics/AbcSize
+
+  def count_all_duplicates(array_storage)
+    array_storage.each_with_object(Hash.new(0)) { |word, counts| counts[word] += 1 }
+  end
+
+  def storage_for_all_text(name)
+    text_storage = ''
+    if list_all_rapers.include? name
+      Dir.glob('rap-battles/*') do |file_name|
```
Comment (stahor) [inline 2226/2/versus.rb]: inject

File: 2226/2/versus.rb [hunk 20]
```diff
@@ -0,0 +1,197 @@
+require 'terminal-table'
```
Comment (stahor) [inline 2226/2/versus.rb]: Global comment: isn't it easier to go through all files one time and collect all data u need?

File: 2226/2/versus.rb [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+          text = File.read(file_name)
+          text_storage += text
+        end
+      end
+    end
+    text_storage
+  end
+
+  def the_most_used_words(top_words, name)
+    array_storage = storage_for_all_text(name).gsub(/,|'/, ' ').split.reject! { |x| x.length < 4 }
```
Comment (stahor) [inline 2226/2/versus.rb]: prepositions filter? what about дом, стул, окно and many other short words

File: 2226/2/versus.rb [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+    array_storage.map(&:capitalize!)
```
Comment (stahor) [inline 2226/2/versus.rb]: better use `map!` and `capitalize`(without bang) it's easier to mutate one object than thousand

File: 2226/2/versus.rb [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+    hash_storage = count_all_duplicates(array_storage)
+    hash_storage = hash_storage.sort_by { |_key, value| value }.reverse.to_h
+    hash_storage.first(top_words.to_i)
+  end
+
+  # rubocop:disable Metrics/AbcSize
+  def print_top_words(top_words, name)
+    if list_all_rapers.include? name
```
Comment (stahor) [inline 2226/2/versus.rb]: parentheses

File: 2226/2/versus.rb [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+    average.round(2)
+  end
+
+  def number_of_words_in_rounds(name)
+    number_of_words_in_rounds = 0
+
+    Dir.glob('rap-battles/*') do |file_name|
+      if file_name.match(name)
+        text = File.read(file_name)
+        number_of_words_in_rounds += text.split(' ').length
```
Comment (stahor) [inline 2226/2/versus.rb]: ''слово" and "слово," are 2 different words in ur algo

File: 2226/2/versus.rb (hunk 8 above)
Comment (andreydro) [reply 2226/2/versus.rb]: I push elements to array. I can use push, <<, insert methods. Can't use += in here

File: 2226/2/versus.rb [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+    sorted_storage = sort_and_reverse_hash(unsorted_storage)
+    sorted_storage.first(number_of_rapers.to_i)
+  end
+
+  def make_table(number_of_rapers)
+    filtered_data = statistics(number_of_rapers).map { |elem| [elem[0], elem[1]] }.to_h
+    rows = []
+    filtered_data.each do |k, v|
+      rows << [
+        k.to_s, "#{v[0]} баттлов", "#{v[1]} нецензурных слов",
```
Comment (stahor) [inline 2226/2/versus.rb]: u can use gem Russian to format this better (pluralize)

File: 2226/2/versus.rb [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+    filter_rapers.flatten!.uniq!
+  end
+
+  # :reek:DuplicateMethodCall
+  def number_of_battles(name)
+    battles_storage = []
+
+    all_data_storage.each do |file|
+      battles_storage << file.first.match(name)
+      battles_storage.reject!(&:nil?)
```
Comment (stahor) [inline 2226/2/versus.rb]: do this after collecting all data not at every step

File: 2226/2/versus.rb [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+
+  def list_all_rapers
+    filter_rapers.flatten!.uniq!
+  end
+
+  # :reek:DuplicateMethodCall
+  def number_of_battles(name)
+    battles_storage = []
+
+    all_data_storage.each do |file|
```
Comment (stahor) [inline 2226/2/versus.rb]: each_with_object

File: 2226/2/versus.rb (hunk 25 above)
Comment (stahor) [inline 2226/2/versus.rb]:
```
[
  value1,
  value2,
  ...
]
```
It's much more readable

File: 2226/2/versus.rb [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+  def make_table(number_of_rapers)
+    filtered_data = statistics(number_of_rapers).map { |elem| [elem[0], elem[1]] }.to_h
+    rows = []
+    filtered_data.each do |k, v|
+      rows << [
+        k.to_s, "#{v[0]} баттлов", "#{v[1]} нецензурных слов",
+        "#{v[2]} слова на баттл", "#{v[3]} слова в раунде"
+      ]
+    end
+    Terminal::Table.new rows: rows
```
Comment (stahor) [inline 2226/2/versus.rb]: parentheses

File: 2226/2/versus.rb [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+
+  def average_number_swearing_words_in_battle(name)
+    average = number_of_swear_words(name).fdiv(number_of_battles(name))
+    average.round(2)
+  end
+
+  def number_of_words_in_rounds(name)
+    number_of_words_in_rounds = 0
+
+    all_data_storage.inject([]) do |_, file|
```
Comment (stahor) [inline 2226/2/versus.rb]: xD
just count with block

File: 2226/2/versus.rb [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+class Raper
+  SWEAR_WORDS_ARRAY = %w[
+    блять блядь бляди ебать ёб бля блядина блядский блядистость блядогон
+    блядословник блядский блядство въебаться взбляд впиздячил выблядовал выблядок
+    выебон выёбывается глупизди доебался ебанёшься ебанул ебашит ёбнул ебало
+    ебанулся ебнул жидоёб жидоёбский заёб изъебнулся заебал заебись козлоёб
+    козлоёбище хуй пизда пиздец \* сука
+  ].freeze
+
+  PRONOUNS_ARRAY = %w[
```
Comment (stahor) [inline 2226/2/versus.rb]: move into config file

File: 2226/2/versus.rb [hunk 31]
```diff
@@ -0,0 +1,199 @@
...
+require 'optparse'
+# :reek:RepeatedConditional
+# :reek:TooManyStatements
+# :reek:UncommunicativeVariableName
+# :reek:UtilityFunction
+# :reek:FeatureEnvy
+# :reek:TooManyMethods
+# rubocop:disable Metrics/ClassLength
+class Raper
+  SWEAR_WORDS_ARRAY = %w[
```
Comment (stahor) [inline 2226/2/versus.rb]: to config

File: 2226/2/versus.rb [hunk 32]
```diff
@@ -0,0 +1,199 @@
...
+      number_of_words_in_rounds += file.last.split(/\W/).count if file.first.match(name)
+    end
+    number_of_words_in_rounds
+  end
+
+  # :reek:DuplicateMethodCall
+  def number_of_rounds(name)
+    round_words = []
+
+    all_data_storage.inject([]) do |_, file|
```
Comment (stahor) [inline 2226/2/versus.rb]: ><
Read docs for inject once again

File: 2226/2/versus.rb [hunk 33]
```diff
@@ -0,0 +1,199 @@
...
+    end
+    battles_storage << 'at least one battle' if battles_storage.size.zero?
+
+    battles_storage.size / 2
+  end
+
+  def number_of_swear_words(name)
+    swear_words_storage = []
+
+    all_data_storage.inject([]) do |_, file|
```
Comment (stahor) [inline 2226/2/versus.rb]: wtf. Use first argument of inject instead of creating storage outside

File: 2226/2/versus.rb (hunk 25 above)
Comment (andreydro) [reply 2226/2/versus.rb]: I need to reduce the number of lines in method, so a followed another syntax style - rubocop respond that it ok
```
[value1,
 vlaue2,
...]
```

File: 2226/2/config.rb [hunk 34]
```diff
@@ -0,0 +1,12 @@
+SWEAR_WORDS_ARRAY = %w[
```
Comment (stahor) [inline 2226/2/config.rb]: better use YAML

File: 2226/2/versus.rb [hunk 35]
```diff
@@ -0,0 +1,171 @@
+require 'terminal-table'
+require 'fuzzy_match'
+require 'optparse'
+require 'russian'
+require_relative 'config'
+# :reek:RepeatedConditional
```
Comment (stahor) [inline 2226/2/versus.rb]: ??? Do u really need this?

File: 2226/2/versus.rb [hunk 36]
```diff
@@ -0,0 +1,192 @@
...
+    filter_rapers.flatten.uniq
+  end
+
+  def self.filter_rapers_from_the_list(file_name)
+    filter = file_name.sub(/ротив_/, '').sub(/_\(.*/, '')
+    filter.sub(%r/^rap-battles\/_/, '').sub(/_$/, '')
+  end
+
+  def self.filter_rapers
+    show_all_data.inject([]) do |array, (file_name, _file_content)|
```
Comment (stahor) [inline 2226/2/versus.rb]: `show_all_data.sum { |file_name, _|  filter_rapers_from_the_list(file_name).split('_п') }`

File: 2226/2/versus.rb [hunk 37]
```diff
@@ -0,0 +1,192 @@
...
+    end
+  end
+end
+
+# This class is used for single raper methods
+class Raper
+  SWEAR_WORDS = YAML.load_file('config.yml')['SWEAR_WORDS_ARRAY'].join('|')
+
+  def self.avg_number_of_battles(name)
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, _file_content)|
```
Comment (stahor) [inline 2226/2/versus.rb]: sum

File: 2226/2/versus.rb [hunk 38]
```diff
@@ -0,0 +1,192 @@
...
+    end
+  end
+
+  def self.number_of_battles(name)
+    avg_number_of_battles(name) / 2
+  end
+
+  def self.number_of_swear_words(name)
+    real_name = name
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/versus.rb]: sum

File: 2226/2/versus.rb [hunk 39]
```diff
@@ -0,0 +1,192 @@
...
+      counter += file_content.scan(/#{SWEAR_WORDS}/).size if file_name.match(real_name)
+      counter
+    end
+  end
+
+  def self.average_number_swearing_words_in_battle(name)
+    average = number_of_swear_words(name).fdiv(number_of_battles(name))
+    average.round(2)
```
Comment (stahor) [inline 2226/2/versus.rb]: do not mix data processing and presentation. This rounding should be somewhere in table drawing method.

File: 2226/2/versus.rb [hunk 40]
```diff
@@ -0,0 +1,192 @@
...
+
+  def self.number_of_words_in_rounds(name)
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, file_content)|
+      counter += file_content.split(/\W/).size if file_name.match(name)
+      counter
+    end
+  end
+
+  def self.number_of_rounds(name)
+    DataStorage.show_all_data.inject(1) do |counter, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/versus.rb]: sum

File: 2226/2/versus.rb [hunk 41]
```diff
@@ -0,0 +1,192 @@
...
+  end
+
+  def statistics(number_of_rapers)
+    sort_and_reverse_hash(PrintTable.sort_storage).first(number_of_rapers.to_i).map { |elem| [elem[0], elem[1]] }.to_h
+  end
+
+  def self.sort_storage
+    DataStorage.list_all_rapers.each_with_object({}) do |raper, hash|
+      hash[raper] = [
+        Raper.number_of_battles(raper), Raper.number_of_swear_words(raper),
```
Comment (stahor) [inline 2226/2/versus.rb]: each element to separate line

File: 2226/2/versus.rb [hunk 42]
```diff
@@ -0,0 +1,192 @@
...
+        Raper.average_number_swearing_words_in_battle(raper), Raper.average_number_words_in_round(raper)
+      ]
+    end
+  end
+
+  # :reek:DuplicateMethodCall # ridiculous warning
+  def make_table(number_of_rapers)
+    rows = []
+    statistics(number_of_rapers).each do |key, value|
```
Comment (stahor) [inline 2226/2/versus.rb]: `row = collection.map do ...`

File: 2226/2/versus.rb [hunk 43]
```diff
@@ -0,0 +1,192 @@
+require 'terminal-table'
+require 'fuzzy_match'
+require 'optparse'
+require 'russian'
+require 'yaml'
+
+# This class is used for storing data
+class DataStorage
```
Comment (stahor) [inline 2226/2/versus.rb]: Use singleton

File: 2226/2/versus.rb [hunk 44]
```diff
@@ -0,0 +1,192 @@
...
+  def self.show_all_data
+    Dir.glob('rap-battles/*').each_with_object({}) do |file_name, hash|
+      hash[file_name] = File.read(file_name)
+    end
+  end
+
+  def self.list_all_rapers
+    filter_rapers.flatten.uniq
```
Comment (stahor) [inline 2226/2/versus.rb]: use memoization `||=`

File: 2226/2/versus.rb [hunk 45]
```diff
@@ -0,0 +1,192 @@
...
+    number_of_words_in_rounds(name) / number_of_rounds(name)
+  end
+end
+
+# :reek:FeatureEnvy # ridiculous warning, basic terminal-table config offer such code
+# This class is used for printing table to the terminal
+class PrintTable
+  def initialize(top_bad_words)
+    @top_bad_words = top_bad_words
+    puts make_table(@top_bad_words)
```
Comment (stahor) [inline 2226/2/versus.rb]: into separate method

File: 2226/2/versus.rb [hunk 46]
```diff
@@ -0,0 +1,192 @@
...
+               "#{value[1]} нецензурных слов",
+               "#{value[2]} слова на баттл",
+               "#{value[3]} слова в раунде"]
+    end
+    Terminal::Table.new(rows: rows)
+  end
+
+  private
+
+  # :reek:UtilityFunction
```
Comment (stahor) [inline 2226/2/versus.rb]: into separate module

File: 2226/2/versus.rb [hunk 47]
```diff
@@ -0,0 +1,192 @@
...
+      end
+    else
+      puts "Репер #{name} не известен мне. Зато мне известны: "
+      puts DataStorage.list_all_rapers
+    end
+  end
+
+  private
+
+  # :reek:UtilityFunction
```
Comment (stahor) [inline 2226/2/versus.rb]: remove

File: 2226/2/versus.rb [hunk 48]
```diff
@@ -0,0 +1,192 @@
...
+  def count_all_duplicates(array_storage)
+    array_storage.each_with_object(Hash.new(0)) { |word, counts| counts[word] += 1 }
+  end
+
+  # :reek:UtilityFunction
```
Comment (stahor) [inline 2226/2/versus.rb]: remove

File: 2226/2/versus.rb (hunk 37 above)
Comment (drazdou1pt) [reply 2226/2/versus.rb]: Can't use sum and count because of if operator

File: 2226/2/versus.rb (hunk 38 above)
Comment (drazdou1pt) [reply 2226/2/versus.rb]: Can't use sum and count because of if operator

File: 2226/2/versus.rb (hunk 37 above)
Comment (andreydro) [reply 2226/2/versus.rb]: Can't use sum and count because of if operator

File: 2226/2/data_storage.rb [hunk 49]
```diff
@@ -0,0 +1,8 @@
+# This module is used for storing data
+module DataStorage
```
Comment (stahor) [inline 2226/2/data_storage.rb]: why module? What if u need more than one storage? 
Create an instance of storage class instead

File: 2226/2/print_table.rb [hunk 50]
```diff
@@ -0,0 +1,21 @@
+# This class is used for printing table to the terminal
+class PrintTable
```
Comment (stahor) [inline 2226/2/print_table.rb]: class names are nouns generally. `TablePrinte`r sounds much better

Comment (stahor) [reply 2226/2/print_table.rb]: this is basically a class that implements level 1 logic but according to its title, the only thing it does is table printing...

File: 2226/2/raper_decorator.rb [hunk 51]
```diff
@@ -0,0 +1,19 @@
+# Class for raper decorator
+class RaperDecorator < Raper
```
Comment (stahor) [inline 2226/2/raper_decorator.rb]: it's more like raper row presenter

File: 2226/2/raper.rb [hunk 52]
```diff
@@ -0,0 +1,50 @@
...
+      file_name.scan(@raper_name).size
+    end
+  end
+
+  def number_of_battles
+    avg_number_of_battles / 2
+  end
+
+  def number_of_swear_words
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/raper.rb]: `sum` with block

File: 2226/2/raper.rb [hunk 53]
```diff
@@ -0,0 +1,50 @@
...
+  def number_of_words_in_rounds
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, file_content)|
+      counter += file_content.split(/ [а-яА-ЯёЁ*]+/).size if file_name.match(@raper_name)
+      counter
+    end
+  end
+
+  def number_of_rounds
+    real_raper_name = @raper_name
+    DataStorage.show_all_data.inject(1) do |counter, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/raper.rb]: this class depends on data storage very much. Isn't it better to let rapper know what battles belong to him from the very beginning and reuse this data?

File: 2226/2/raper.rb [hunk 54]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def avg_number_of_battles
+    DataStorage.show_all_data.sum do |file_name, _|
+      file_name.scan(@raper_name).size
+    end
+  end
+
+  def number_of_battles
+    avg_number_of_battles / 2
```
Comment (stahor) [inline 2226/2/raper.rb]: are u sure that there is a pair for every battle?

File: 2226/2/raper.rb [hunk 55]
```diff
@@ -0,0 +1,50 @@
...
+class Raper
+  attr_reader :raper_name
+
+  SWEAR_WORDS = YAML.load_file('config.yml')['SWEAR_WORDS_ARRAY'].join('|')
+
+  def initialize(raper_name)
+    @raper_name = raper_name
+  end
+
+  def avg_number_of_battles
```
Comment (stahor) [inline 2226/2/raper.rb]: well, this method should be private I guess

File: 2226/2/raper.rb [hunk 56]
```diff
@@ -0,0 +1,50 @@
...
+      counter
+    end
+  end
+
+  def average_number_swearing_words_in_battle
+    number_of_swear_words.fdiv(number_of_battles)
+  end
+
+  def number_of_words_in_rounds
+    DataStorage.show_all_data.inject(0) do |counter, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/raper.rb]: sum

File: 2226/2/raper.rb [hunk 57]
```diff
@@ -0,0 +1,50 @@
...
+      counter += file_content.split(/ [а-яА-ЯёЁ*]+/).size if file_name.match(@raper_name)
```
Comment (stahor) [inline 2226/2/raper.rb]: `file_name.match(@raper_name) `
is repeated everywhere. It has to be a separate method probably. But actually, you don't need it only once.

File: 2226/2/utility_methods.rb [hunk 58]
```diff
@@ -0,0 +1,21 @@
+# This module is used for utility methods
+module UtilityMethods
+  PRONOUNS = YAML.load_file('config.yml')['PRONOUNS_ARRAY'].join('|')
+
+  def self.count_all_duplicates(array_storage)
```
Comment (stahor) [inline 2226/2/utility_methods.rb]: [read this](https://ruby-doc.org/core-2.2.0/Module.html#module_function-method)

File: 2226/2/utility_methods.rb [hunk 59]
```diff
@@ -0,0 +1,21 @@
...
+    array_storage.each_with_object(Hash.new(0)) { |word, counts| counts[word] += 1 }
+  end
+
+  def self.storage_for_all_text(name)
+    DataStorage.show_all_data.inject('') do |string, (file_name, file_content)|
```
Comment (stahor) [inline 2226/2/utility_methods.rb]: use each_with_object instead and remove last `string` in block

File: 2226/2/utility_methods.rb [hunk 60]
```diff
@@ -0,0 +1,21 @@
+# This module is used for utility methods
+module UtilityMethods
+  PRONOUNS = YAML.load_file('config.yml')['PRONOUNS_ARRAY'].join('|')
+
+  def self.count_all_duplicates(array_storage)
+    array_storage.each_with_object(Hash.new(0)) { |word, counts| counts[word] += 1 }
+  end
+
+  def self.storage_for_all_text(name)
```
Comment (stahor) [inline 2226/2/utility_methods.rb]: no needed if u keep files inside raper instance

File: 2226/2/utility_methods.rb [hunk 61]
```diff
@@ -0,0 +1,21 @@
...
+    DataStorage.show_all_data.inject('') do |string, (file_name, file_content)|
+      string += file_content if file_name.match(name)
+      string
+    end
+  end
+
+  def self.the_most_used_words(top_words, name)
```
Comment (stahor) [inline 2226/2/utility_methods.rb]: this is part of application logic. Why is this among utility functions?

File: 2226/2/print_top_words.rb [hunk 62]
```diff
@@ -0,0 +1,20 @@
...
+  def initialize(top_words, raper_name)
+    @top_words = top_words
+    @name = raper_name
+    @rapers_storage = ListOfRapers.list_all_rapers
+    print_top_words(@top_words, @name)
+  end
+
+  def print_top_words(top_words, name)
+    if @rapers_storage.include?(name)
+      UtilityMethods.the_most_used_words(top_words, name).each do |elem|
```
Comment (stahor) [inline 2226/2/print_top_words.rb]: so the whole logic of level 2 is inside utility method...

File: 2226/2/helper.rb [hunk 63]
```diff
@@ -0,0 +1,14 @@
+# Module for help
+module Helper
+  PRONOUNS = YAML.load_file('config.yml')['PRONOUNS_ARRAY'].join('|')
+
+  def duplicates_counter(array_storage)
```
Comment (stahor) [inline 2226/2/helper.rb]: it's a histogramm actually

File: 2226/2/data_storage.rb [hunk 64]
```diff
@@ -0,0 +1,8 @@
+# This class sis used for storing data
+class DataStorage
+  def self.show_all_data
```
Comment (stahor) [inline 2226/2/data_storage.rb]: create an instance and use it. And you can pass the path to the folder into `#new`.

File: 2226/2/battle.rb [hunk 65]
```diff
@@ -0,0 +1,62 @@
+# Class to battles of raper
+class Battle
+  include Helper
+
+  SWEAR_WORDS = YAML.load_file('config.yml')['SWEAR_WORDS_ARRAY'].join('|')
+
+  def initialize(raper_name)
+    @raper_name = raper_name
+    @all_data_hash = DataStorage.show_all_data
```
Comment (stahor) [inline 2226/2/battle.rb]: why battle depends on storage?

File: 2226/2/battle.rb [hunk 66]
```diff
@@ -0,0 +1,62 @@
...
+    @data_for_one_raper = filter_one_raper_data
```
Comment (stahor) [inline 2226/2/battle.rb]: move to separate method with memoization

File: 2226/2/raper.rb [hunk 67]
```diff
@@ -0,0 +1,38 @@
+# This class is used for single raper methods
+class Raper
+  attr_reader :raper_name
+
+  def initialize(raper_name)
+    @raper_name = raper_name
+  end
+
+  def number_of_battles
+    Battle.new(@raper_name).number_of_battles
```
Comment (stahor) [inline 2226/2/raper.rb]: Use delegate pattern for things like this. But Actually what u are doing is totally wrong. You can remove rapper class from this program and use Battle class only.

File: 2226/2/table_printer.rb [hunk 68]
```diff
@@ -0,0 +1,21 @@
+# This class is used for printing table to the terminal
+class TablePrinter
```
Comment (stahor) [inline 2226/2/table_printer.rb]: This class still contains a lot of logic program logic but it's name is just table printer. Does it print HTML tables code?

File: 2226/2/battle.rb [hunk 69]
```diff
@@ -0,0 +1,27 @@
...
+    @text.scan(/#{SWEAR_WORDS}/).size
+  end
+
+  def number_of_words_in_rounds
+    @text.split(/ [а-яА-ЯёЁ*]+/).size
+  end
+
+  def number_of_rounds
+    rounds = @text.scan(/Раунд \w/).size
+    return rounds unless rounds.zero?
```
Comment (stahor) [inline 2226/2/battle.rb]: `rounds.zero? ? 1: rounds`

File: 2226/2/data_storage.rb [hunk 70]
```diff
@@ -0,0 +1,12 @@
+# This class sis used for storing data
+class DataStorage
+  def new
```
Comment (stahor) [inline 2226/2/data_storage.rb]: O_O

File: 2226/2/list_of_rapers.rb [hunk 71]
```diff
@@ -0,0 +1,13 @@
+# This module storing rapers
+module ListOfRapers
```
Comment (stahor) [inline 2226/2/list_of_rapers.rb]: this all should be part of data store.

File: 2226/2/raper.rb [hunk 72]
```diff
@@ -0,0 +1,57 @@
...
+
+  def the_most_used_words(top_words)
+    array_storage = data_for_one_raper.values.join(' ').gsub(/,|'/, ' ').split
+    array_storage = delete_prons(array_storage)
+    count_dupls(array_storage).sort_by { |_, value| value }.reverse.to_h.first(top_words.to_i)
+  end
+
+  private
+
+  def count_dupls(array_storage)
```
Comment (stahor) [inline 2226/2/raper.rb]: can u use `duplicates_counter(array_storage)` directly instead of this method?

File: 2226/2/battle.rb [hunk 73]
```diff
@@ -0,0 +1,26 @@
+# This class is used to process raper data
+class Battle
+  SWEAR_WORDS = YAML.load_file('config.yml')['SWEAR_WORDS_ARRAY'].join('|')
+
+  def initialize(name_of_battle, text)
+    @name_of_battle = name_of_battle
```
Comment (stahor) [inline 2226/2/battle.rb]: this variable is never used

**Pull request merged**